### PR TITLE
[Feature] Add docker files

### PR DESCRIPTION
**Pull request by AllentDan**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

To provide docker files for building stable env for customers.

## Modification

Added docker/*

## BC-breaking (Optional)

No

### CODE REVIEW

File: docker/CPU/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    git clone https://github.com/Tencent/ncnn.git &&\
+    cd ncnn &&\
+    export NCNN_DIR=$(pwd) &&\
+    git submodule update --init &&\
+    mkdir -p build && cd build &&\
+    cmake -DNCNN_VULKAN=OFF -DNCNN_SYSTEM_GLSLANG=ON -DNCNN_BUILD_EXAMPLES=ON -DNCNN_PYTHON=ON -DNCNN_BUILD_TOOLS=ON -DNCNN_BUILD_BENCHMARK=ON -DNCNN_BUILD_TESTS=ON .. &&\
+    make install &&\
+    cd ${NCNN_DIR} # To NCNN root directory &&\
+    cd python &&\
+    pip install -e .
```
[the remark this answers is not in the document]
Comment (AllentDan) [reply docker/CPU/Dockerfile:71]: I chose to use pip for all cases. There is no python2 installed yet

File: docker/CPU/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    rm ~/miniconda.sh && \
+    /opt/conda/bin/conda install -y python=${PYTHON_VERSION} conda-build pyyaml numpy ipython cython typing typing_extensions mkl mkl-include ninja && \
+    /opt/conda/bin/conda clean -ya
+
+### pytorch
+RUN /opt/conda/bin/pip install torch==${TORCH_VERSION}+cpu torchvision==${TORCHVISION_VERSION}+cpu -f https://download.pytorch.org/whl/cpu/torch_stable.html
+ENV PATH /opt/conda/bin:$PATH
+
+### install open-mim
+RUN /opt/conda/bin/pip install mmcv-full -f https://download.openmmlab.com/mmcv/dist/cpu/torch${TORCH_VERSION}/index.html
```
Comment (RunningLeon) [inline docker/CPU/Dockerfile:35]: may add a mmcv version?

File: docker/GPU/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    /opt/conda/bin/conda install -y python=${PYTHON_VERSION} conda-build pyyaml numpy ipython cython typing typing_extensions mkl mkl-include ninja && \
+    /opt/conda/bin/conda clean -ya
+
+### pytorch
+RUN /opt/conda/bin/conda install pytorch==${TORCH_VERSION} torchvision==${TORCHVISION_VERSION} cudatoolkit=${CUDA} -c pytorch
+ENV PATH /opt/conda/bin:$PATH
+
+RUN ls /opt/conda/bin/
+### install open-mim
+RUN /opt/conda/bin/pip install mmcv-full -f https://download.openmmlab.com/mmcv/dist/cu${CUDA//./}/torch${TORCH_VERSION}/index.html
```
Comment (RunningLeon) [inline docker/GPU/Dockerfile:29]: may need to set a mmcv version

File: docker/GPU/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+### git mmdetection
+RUN git clone --depth=1 https://github.com/open-mmlab/mmdetection.git /root/space/mmdetection
+
+### install mmdetection
+RUN cd /root/space/mmdetection &&\
+    pip3 install -r requirements.txt &&\
```
Comment (RunningLeon) [inline docker/GPU/Dockerfile:36]: should we use pip3 and python3 here?

File: docker/GPU/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    python3 setup.py develop
+
+WORKDIR /root/workspace
+RUN git clone https://github.com/open-mmlab/mmclassification
+
+
+### get onnxruntime
+RUN wget https://github.com/microsoft/onnxruntime/releases/download/v${ONNXRUNTIME_VERSION}/onnxruntime-linux-x64-${ONNXRUNTIME_VERSION}.tgz \
+    && tar -zxvf onnxruntime-linux-x64-${ONNXRUNTIME_VERSION}.tgz &&\
+    pip install onnxruntime-gpu
```
Comment (RunningLeon) [inline docker/GPU/Dockerfile:46]: may need tp set the version of onnxruntime-gpu

File: docker/GPU/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+### build ppl.nn
+WORKDIR /root/workspace
+RUN git clone https://github.com/openppl-public/ppl.nn &&\
+    cd ppl.nn &&\
+    ./build.sh -DPPLNN_ENABLE_PYTHON_API=ON -DHPCC_USE_X86_64=ON -DHPCC_USE_CUDA=ON &&\
+    rm -rf /tmp/pyppl-package &&\
+    cp -r python/package /tmp/pyppl-package &&\
+    cp -r pplnn-build/install/lib/pyppl/* /tmp/pyppl-package/pyppl &&\
+    cd /tmp/pyppl-package &&\
+    pip3 install .
```
Comment (RunningLeon) [inline docker/GPU/Dockerfile:72]: pip3?

File: docker/CPU/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+ARG ONNXRUNTIME_VERSION=1.8.1
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt-get update && apt-get install -y --no-install-recommends \
+        ca-certificates \
+        build-essential \
+        libjpeg-dev \
+        libpng-dev \
+        ccache \
+        cmake \
+        gcc \
```
[the remark this answers is not in the document]
Comment (AllentDan) [reply docker/CPU/Dockerfile:18]: No need to specify the versions when it is Ubuntu 18.04

File: docker/README.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
+## Docker usage
+
+We provide two dockerfiles for CPU and GPU respectively. For CPU users, we install MMDeploy with ONNXRuntime, ncnn and OpenVINO backends. For GPU users, we install MMDeploy with TensorRT backend. Besides, users can install mmdeploy with different versions when building the docker image.
+
+### Build docker image
+
+For CPU users, we can build the docker image with the latest MMDeploy through:
+```
+cd mmdeploy
+docker build docker/CPU/ -t mmdeploy:master
```
Comment (fbagci) [inline docker/README.md:10]: Maybe image tags could be?
For GPU: mmdeploy:master -> mmdeploy:gpu-master
For CPU: mmdeploy:master -> mmdeploy:cpu-master

Comment (AllentDan) [reply docker/README.md:10]: Thanks for the suggestion. Will use `mmdeploy:master-gpu` :smiley:

File: docker/CPU/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    -DCMAKE_CXX_COMPILER=g++-7 \
+    -DONNXRUNTIME_DIR=${ONNXRUNTIME_DIR} \
+    -Dncnn_DIR=/root/workspace/ncnn/build/install/lib/cmake/ncnn \
+    -DInferenceEngine_DIR=/opt/intel/openvino/deployment_tools/inference_engine/share \
+    -DMMDEPLOY_TARGET_DEVICES=cpu \
+    -DMMDEPLOY_BUILD_SDK_PYTHON_API=ON \
+    -DMMDEPLOY_TARGET_BACKENDS="ort;ncnn;openvino" \
+    -DMMDEPLOY_CODEBASES=all &&\
+    cmake --build . -- -j$(nproc) && cmake --install . &&\
+    if [ -z ${VERSION} ] ; then echo "Build MMDeploy master for CPU devices succeed!" ; else echo "Build MMDeploy version v${VERSION} for CPU devices succeed!" ; fi
```
Comment (RunningLeon) [inline docker/CPU/Dockerfile:101]: Built xxx successfully.

File: docker/CPU/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    rm ~/miniconda.sh && \
+    /opt/conda/bin/conda install -y python=${PYTHON_VERSION} conda-build pyyaml numpy ipython cython typing typing_extensions mkl mkl-include ninja && \
+    /opt/conda/bin/conda clean -ya
+
+### pytorch
+RUN /opt/conda/bin/pip install torch==${TORCH_VERSION}+cpu torchvision==${TORCHVISION_VERSION}+cpu -f https://download.pytorch.org/whl/cpu/torch_stable.html
+ENV PATH /opt/conda/bin:$PATH
+
+### install open-mim
+RUN /opt/conda/bin/pip install mmcv-full==1.4.0 -f https://download.openmmlab.com/mmcv/dist/cpu/torch${TORCH_VERSION}/index.html
```
Comment (RunningLeon) [inline docker/CPU/Dockerfile:38]: Could we create a variable to the version of `mmcv-full` at the beginning?

File: docker/GPU/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    rm ~/miniconda.sh && \
+    /opt/conda/bin/conda install -y python=${PYTHON_VERSION} conda-build pyyaml numpy ipython cython typing typing_extensions mkl mkl-include ninja && \
+    /opt/conda/bin/conda clean -ya
+
+### pytorch
+RUN /opt/conda/bin/conda install pytorch==${TORCH_VERSION} torchvision==${TORCHVISION_VERSION} cudatoolkit=${CUDA} -c pytorch
+ENV PATH /opt/conda/bin:$PATH
+
+### install mmcv-full
+RUN /opt/conda/bin/pip install mmcv-full==1.4.0 -f https://download.openmmlab.com/mmcv/dist/cu${CUDA//./}/torch${TORCH_VERSION}/index.html &&\
```
Comment (RunningLeon) [inline docker/GPU/Dockerfile:29]: Could we create a variable to the version of `mmcv-full` at the beginning?

**Comment by lvhan028** [comment]
Error occurred when executing a demo `object_detection`
```bash
./object_detection cuda <path/of/sdk/model> <path/of/an/image>
```
Error log: 
```
[resize.cu:998] CUDA error: the provided PTX was compiled with an unsupported toolchain.
```

File: docs/en/tutorials/how_to_use_docker.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+### Run docker container
+
+After building the docker image succeed, we can use `docker run` to launch the docker service. GPU docker image for example:
+```
+docker run --gpus all -it -p 8080:8081 mmdeploy:master-gpu
+```
+
+### AFQs
```
Comment (RunningLeon) [inline docs/en/tutorials/how_to_use_docker.md:31]: FAQs

Comment (AllentDan) [reply docs/en/tutorials/how_to_use_docker.md:31]: 😅

File: docker/GPU/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    cd build &&\
+    cmake -DMMDEPLOY_TARGET_BACKENDS=ort .. &&\
+    make -j$(nproc) &&\
+    cmake -DMMDEPLOY_TARGET_BACKENDS=trt .. &&\
+    make -j$(nproc) &&\
+    cd .. &&\
+    pip install -e .
+
+### build sdk
+RUN git clone https://github.com/openppl-public/ppl.cv.git &&\
```
Comment (RunningLeon) [inline docker/GPU/Dockerfile:69]: should install pplnn at first.

Comment (AllentDan) [reply docker/GPU/Dockerfile:69]: Seems no need to to that as pplcv is independent from pplnn

File: docs/en/tutorials/how_to_use_docker.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
+## Docker usage
+
+We provide two dockerfiles for CPU and GPU respectively. For CPU users, we install MMDeploy with ONNXRuntime, ncnn and OpenVINO backends. For GPU users, we install MMDeploy with TensorRT backend. Besides, users can install mmdeploy with different versions when building the docker image.
```
Comment (RunningLeon) [inline docs/en/tutorials/how_to_use_docker.md:3]: Why not include pplnn backend in GPU docker?

Comment (AllentDan) [reply docs/en/tutorials/how_to_use_docker.md:3]: Explained off-line during the meeting. 
1. ppl.nn always failed during building, as there are too much dependencies to be downloaded.
2. the latest ppl.nn can't be used when building SDK

**Review by lvhan028 — APPROVED** [review]
LGTM

**Comment by RunningLeon** [comment]
### script
```
sudo docker run --gpus all -it -p 8080:8081 -v ~/projects:/projects mmdeploy:master-gpu
```
### failed message
```
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
ERRO[0000] error waiting for container: context canceled 
```
### solved by [link](https://github.com/NVIDIA/nvidia-docker/issues/1034#issuecomment-608401190)
```
# Add the package repositories
distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add -
curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list

sudo apt-get update && sudo apt-get install -y nvidia-container-toolkit
sudo systemctl restart docker
```

**Comment by RunningLeon** [comment]
@AllentDan May need to fix conflicts.

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**